### PR TITLE
[New Rule] GCP Kubernetes Secrets Enumeration

### LINKED ISSUE #1188: [New Rule] Kubernetes Secrets Enumeration
## Description

Kubernetes is an attacking vector that current there is no detection rules for. During my initial analysis I was able to create a number of KQL queries that can detect the following activity:

The purpose of this rule is to detect a user enumerating kubernetes secrets 

### Target indexes

`filebeat-*` with `gcp` module having `audit` configured.

### Target Operating Systems

Kubernetes

### Platforms

Google Kubernetes Engine

### Tested ECS Version
1.8.0

### Query

In this query I'm excluding any enumeration done by `system*` which could include the following:
- `system:serviceaccount:elastic-system:elastic-operator`
- `system:apiserver`
- `system:serviceaccount:default:ingress-controller-ingress-nginx-admission`
- `system:serviceaccount:kube-system:generic-garbage-collector`

<details>
<summary>Query</summary>

```
"query": {
    "bool": {
      "must": [],
      "filter": [
        {
          "bool": {
            "must_not": {
              "bool": {
                "should": [
                  {
                    "query_string": {
                      "fields": [
                        "gcp.audit.authentication_info.principal_email"
                      ],
                      "query": "system*"
                    }
                  }
                ],
                "minimum_should_match": 1
              }
            }
          }
        },
        {
          "range": {
            "@timestamp": {
              "gte": "2021-05-04T13:29:16.495Z",
              "lte": "2021-05-05T13:29:16.495Z",
              "format": "strict_date_optional_time"
            }
          }
        },
        {
          "match_phrase": {
            "event.module": "gcp"
          }
        },
        {
          "match_phrase": {
            "gcp.audit.type": "type.googleapis.com/google.cloud.audit.AuditLog"
          }
        },
        {
          "match_phrase": {
            "gcp.audit.method_name": "io.k8s.core.v1.secrets.get"
          }
        }
      ],
      "should": [],
      "must_not": [
        {
          "match_phrase": {
            "gcp.audit.status.message": "PERMISSION_DENIED"
          }
        }
      ]
    }
  }
```

</details>

### Related issues or PRs

Ready to create the necessary detection rule `toml` files

### References

https://cloud.google.com/kubernetes-engine/docs/how-to/audit-logging

### PR DESCRIPTION
**Pull request by austinsonger**
## Issues

Resolves #1188

## Contributor checklist

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/detection-rules/blob/main/CONTRIBUTING.md)?

### CODE REVIEW

File: rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+[metadata]
+creation_date = "2021/06/06"
+maturity = "production"
+updated_date = "2021/06/06"
+
```
Comment (w0rk3r) [inline rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml:5]: Missing Integration

File: rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+[rule]
+author = ["Elastic","Austin Songer"]
```
Comment (w0rk3r) [inline rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml:7]:
```suggestion
author = ["Elastic", "Austin Songer"]
```

File: rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+description = """
+Identifies the unauthorized enumeration of kubernetes secrets by a user.
```
Comment (w0rk3r) [inline rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml:9]: How we differentiate the enumeration from legitimate activity? If we can't, we need to reformulate this.

Comment (w0rk3r) [reply rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml:9]: Also, we need to explain why the this is critical, a little context of what can be done using these secrets would be nice here.

File: rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+from = "now-20m"
+index = ["filebeat-*", "logs-gcp*"]
+language = "kuery"
+license = "Elastic License v2"
+name = "GCP Kubernetes Secrets Enumeration"
+note = """## Config
+
+The GCP Fleet integration, Filebeat module, or similarly structured data is required to be compatible with this rule."""
+references = [
+    "https://cloud.google.com/kubernetes-engine/docs/how-to/audit-logging",
```
Comment (w0rk3r) [inline rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml:20]: To be replaced with a reference more relevant to the attack/activity

File: rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+creation_date = "2021/06/06"
+maturity = "production"
+updated_date = "2021/06/06"
+
+[rule]
+author = ["Elastic","Austin Songer"]
+description = """
+Identifies the unauthorized enumeration of kubernetes secrets by a user.
+"""
+from = "now-20m"
```
Comment (w0rk3r) [inline rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml:11]: We need to specify in which instances this operation would be used by administrators in false positives.

File: rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+]
+risk_score = 47
+rule_id = "90b6cf41-801b-4979-9233-e9a234257cd9"
+severity = "medium"
+tags = ["Elastic", "Cloud", "GCP", "Continuous Monitoring", "SecOps", "Configuration Audit"]
+timestamp_override = "event.ingested"
+type = "query"
+
+query = '''
+event.dataset:(googlecloud.audit or gcp.audit) and event.action:io.k8s.core.v*.secrets.get and event.outcome:success
```
Comment (w0rk3r) [inline rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml:30]: I could not find any reference for this event action on google, can you provide us a sample event?

File: rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+'''
+
+[[rule.threat]]
+framework = "MITRE ATT&CK"
+
+[rule.threat.tactic]
+id = "TA0006"
+name = "Credential Access"
+reference = "https://attack.mitre.org/tactics/TA0006/"
+
```
Comment (w0rk3r) [inline rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml:40]: This one seems to be a good fit here: https://attack.mitre.org/techniques/T1552/007/

**Comment by w0rk3r** [comment]
Closing this one due to lack of activity, feel free to open a new PR if we progress. Thanks!

**Comment by austinsonger** [comment]
Can you not close mine right now. I'm on emergency leave and just don't have time right now.

File: rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml (hunk 6 above)
Comment (austinsonger) [reply rules/integrations/gcp/credential_access_gcp_kubernetes_secrets_enumeration.toml:34]:
```
"query": {
    "bool": {
      "must": [],
      "filter": [
        {
          "bool": {
            "must_not": {
              "bool": {
                "should": [
                  {
                    "query_string": {
                      "fields": [
                        "gcp.audit.authentication_info.principal_email"
                      ],
                      "query": "system*"
                    }
                  }
                ],
                "minimum_should_match": 1
              }
            }
          }
        },
        {
          "range": {
            "@timestamp": {
              "gte": "2021-05-04T13:29:16.495Z",
              "lte": "2021-05-05T13:29:16.495Z",
              "format": "strict_date_optional_time"
            }
          }
        },
        {
          "match_phrase": {
            "event.module": "gcp"
          }
        },
        {
          "match_phrase": {
            "gcp.audit.type": "type.googleapis.com/google.cloud.audit.AuditLog"
          }
        },
        {
          "match_phrase": {
            "gcp.audit.method_name": "io.k8s.core.v1.secrets.get"
          }
        }
      ],
      "should": [],
      "must_not": [
        {
          "match_phrase": {
            "gcp.audit.status.message": "PERMISSION_DENIED"
          }
        }
      ]
    }
  }

```

**Pull request closed without merging**